lib/assert.js

```diff
@@ -285,15 +285,24 @@ function _deepEqual(actual, expected, strict, memos) {
   // Note: this accounts for both named and indexed properties on Arrays.
 
   // Use memos to handle cycles.
-  memos = memos || { actual: [], expected: [] };
-  const actualIndex = memos.actual.indexOf(actual);
-  if (actualIndex !== -1) {
-    if (actualIndex === memos.expected.indexOf(expected)) {
+  if (!memos) {
+    memos = {
+      actual: { map: new Map(), position: 0 },
+      expected: { map: new Map(), position: 0 }
+    };
+  }
+
+  const actualPosition = memos.actual.map.get(actual);
+  if (actualPosition !== undefined) {
+    if (actualPosition === memos.expected.map.get(expected)) {
```

Comment:
I don’t think that actually affects the logic here – if it matters at all, it should only affect cases where non-primitives compare deep-equal to primitives, but that doesn’t seem to happen anyway (e.g. `assert.deepEqual([0], 0)` throws even though `[0] == 0`).